remote-storage/src/test/java/org/apache/kafka/server/log/remote/storage/InmemoryRemoteStorageManagerTest.java

```diff
@@ -0,0 +1,241 @@
+        RemoteLogSegmentMetadata rlsm = createRemoteLogSegmentMetadata();
+        int segSize = 100;
+        LogSegmentData logSegmentData = createLogSegmentData(segSize);
+
+        // Copy the segment
+        rsm.copyLogSegmentData(rlsm, logSegmentData);
+
+        // Check fetch segments with invalid ranges like startPos < endPos
+        Assertions.assertThrows(Exception.class, () -> rsm.fetchLogSegment(rlsm, 2, 1));
+
+        // Check fetch segments with invalid ranges like startPos or endPos as negative.
+        Assertions.assertThrows(Exception.class, () -> rsm.fetchLogSegment(rlsm, -1, 0));
+        Assertions.assertThrows(Exception.class, () -> rsm.fetchLogSegment(rlsm, -2, -1));
+    }
+
+    @Test
+    public void testDeleteSegment() throws Exception {
+        InmemoryRemoteStorageManager rsm = new InmemoryRemoteStorageManager();
+        RemoteLogSegmentMetadata rlsm = createRemoteLogSegmentMetadata();
+        LogSegmentData logSegmentData = createLogSegmentData();
+
+        // Copy a log segment.
+        rsm.copyLogSegmentData(rlsm, logSegmentData);
+
+        // Check that the copied segment exists in rsm and it is same.
+        try (InputStream segmentStream = rsm.fetchLogSegment(rlsm, 0)) {
+            checkContentSame(segmentStream, logSegmentData.logSegment().toPath());
+        }
+
+        // Delete segment and check that it does not exist in RSM.
+        rsm.deleteLogSegmentData(rlsm);
+
+        // Check that the segment data does not exist.
+        Assertions.assertThrows(RemoteResourceNotFoundException.class, () -> rsm.fetchLogSegment(rlsm, 0));
+
+        // Check that the segment data does not exist for range.
+        Assertions.assertThrows(RemoteResourceNotFoundException.class, () -> rsm.fetchLogSegment(rlsm, 0, 1));
+
+        // Check that all the indexes are not found.
+        for (RemoteStorageManager.IndexType indexType : RemoteStorageManager.IndexType.values()) {
+            Assertions.assertThrows(RemoteResourceNotFoundException.class, () -> rsm.fetchIndex(rlsm, indexType));
+        }
+    }
+
+    private void checkContentSame(InputStream segmentStream, Path path) throws IOException {
+        byte[] segmentBytes = Files.readAllBytes(path);
+        ByteBuffer byteBuffer = readAsByteBuffer(segmentStream, segmentBytes.length);
+        Assertions.assertEquals(ByteBuffer.wrap(segmentBytes), byteBuffer);
+    }
+
+    private ByteBuffer readAsByteBuffer(InputStream segmentStream,
+                                        int len) throws IOException {
+        ByteBuffer byteBuffer = ByteBuffer.wrap(new byte[len]);
+        Utils.readFully(segmentStream, byteBuffer);
+        byteBuffer.rewind();
+        return byteBuffer;
+    }
+
+    private LogSegmentData createLogSegmentData() throws Exception {
+        return createLogSegmentData(100);
+    }
+
+    private LogSegmentData createLogSegmentData(int segSize) throws Exception {
+        int prefix = Math.abs(new Random().nextInt());
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [DMI: Random object created and used only once (DMI_RANDOM_USED_ONLY_ONCE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#dmi-random-object-created-and-used-only-once-dmi-random-used-only-once).
This code creates a java.util.Random object, uses it to generate one random number, and then discards the Random object. This produces mediocre quality random numbers and is inefficient. If possible, rewrite the code so that the Random object is created once and saved, and each time a new random number is required invoke a method on the existing Random object to obtain it.
If it is important that the generated Random numbers not be guessable, you *must* not create a new Random for each random number; the values are too easily guessable. You should strongly consider using a java.security.SecureRandom instead (and avoid allocating a new SecureRandom for each random number needed).